torchdata/dataloader2/dataloader2.py

```diff
@@ -130,6 +133,9 @@ def __iter__(self) -> Iterator[T_co]:
         self.valid_iterator_id = 0 if self.valid_iterator_id is None else self.valid_iterator_id + 1
         return DataLoader2Iterator(self, self.valid_iterator_id)
 
+    def __len__(self):
+        return len(self.datapipe)
```

Comment:
I think we still need to rely on `self.datapipe` because RS won't have aware of length of `self.datapipe` before `__iter__` is invoked (as `reading_service.initialize(datapipe)`). And, after `initialize(datapipe)`, the `self.datapipe` should have been modified and the corresponding changes should be reflected at length.

But, this behavior becomes super weird to me. The length varies before and after `__iter__` is invoked.

Reply:
Could we do something like adapting `self.datapipe` when `__len__` is called but `self.datapipe` has not been modified by `self.datapipe`?

Reply:
I am going to change it to `len(self._datapipe_before_reading_service_adapt)` because I am guessing that what you mean. It makes sense since that variable also shouldn't be impacted by ReadingService but it would be impacted by adapters within `__init__`, but that is a little bit dependent on the implementation of ReadingService. I think if you `.apply_sharding` within child process and not the main one, the length should not change.

Relevant issue about what is the proper `__len__` to return: https://github.com/pytorch/data/issues/533